### PR TITLE
add interface to auto clean mysql pendingSegments table

### PR DESCRIPTION
**Pull request by haoxiang47**
For the issue 3565, I finished the interface to delete unused pendingSeglments, add TaskDataSegment.java to struct the task table’s data, add getNotActiveTask and deletePendingSegments in IndexerSQLMetadataStorageCoordinator.java to finish the delete logic, add test in TestIndexerMetadataStorageCoordinator.java.

### CODE REVIEW

**Comment by fjy** [comment]
@KurtYoung can you take a look?

File: api/src/main/java/io/druid/timeline/TaskDataSegment.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
+package io.druid.timeline;
```
Comment (fjy) [inline api/src/main/java/io/druid/timeline/TaskDataSegment.java]: missing license header

Comment (fjy) [inline api/src/main/java/io/druid/timeline/TaskDataSegment.java]: @haoxiang47 there are many whitespace changes, can we please revert them?
https://github.com/druid-io/druid/pull/3831/files?w=1 indicates there's only new code added

File: api/src/main/java/io/druid/timeline/TaskDataSegment.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import org.joda.time.Interval;
+
+import java.util.Map;
+
+/**
+ * Created by haoxiang on 16/11/11.
```
Comment (fjy) [inline api/src/main/java/io/druid/timeline/TaskDataSegment.java]: we generally dont have author info

File: api/src/main/java/io/druid/timeline/TaskDataSegment.java [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    {
+        this.type = type;
+        this.id = id;
+        this.groupId = groupId;
+        this.dataSource = dataSource;
+        this.ioConfig = ioConfig;
+        this.interval = interval;
+    }
+
+    /**
```
Comment (fjy) [inline api/src/main/java/io/druid/timeline/TaskDataSegment.java]: this is a weird comment

File: api/src/main/java/io/druid/timeline/TaskDataSegment.java (hunk 1 above)
Comment (fjy) [inline api/src/main/java/io/druid/timeline/TaskDataSegment.java]: also: https://github.com/druid-io/druid/blob/master/druid_intellij_formatting.xml

File: indexing-service/src/test/java/io/druid/indexing/test/TestIndexerMetadataStorageCoordinator.java [hunk 4]
```diff
@@ -121,6 +124,12 @@ public void deleteSegments(Set<DataSegment> segments)
   }
 
   @Override
+  public List<TaskDataSegment> getNotActiveTask(final Interval interval){
```
Comment (fjy) [inline indexing-service/src/test/java/io/druid/indexing/test/TestIndexerMetadataStorageCoordinator.java]: InactiveTasks*

File: indexing-service/src/test/java/io/druid/indexing/test/TestIndexerMetadataStorageCoordinator.java [hunk 5]
```diff
@@ -39,10 +40,12 @@
   final private Set<DataSegment> published = Sets.newConcurrentHashSet();
   final private Set<DataSegment> nuked = Sets.newConcurrentHashSet();
   final private List<DataSegment> unusedSegments;
+  final private List<TaskDataSegment> taskDataSegments;
```
Comment (fjy) [inline indexing-service/src/test/java/io/druid/indexing/test/TestIndexerMetadataStorageCoordinator.java]: inactiveTaskSegments

File: server/src/main/java/io/druid/indexing/overlord/IndexerMetadataStorageCoordinator.java [hunk 6]
```diff
@@ -141,4 +142,11 @@ SegmentPublishResult announceHistoricalSegments(
    * @return DataSegments which include ONLY data within the requested interval and are not flagged as used. Data segments NOT returned here may include data in the interval
    */
   List<DataSegment> getUnusedSegmentsForInterval(String dataSource, Interval interval);
+
+  /**
+   * Get all task which active property is 0, is used to delete the useless segments in pendingSegment table.
+   * @param interval Filter the tasks to ones that include tasks in this interval exclusively. Start is inclusive, end is exclusive
+   * @return The TaskDataSegment list which used to match the pendingSegment table's payload
+   */
+  List<TaskDataSegment> getNotActiveTask(final Interval interval);
```
Comment (fjy) [inline server/src/main/java/io/druid/indexing/overlord/IndexerMetadataStorageCoordinator.java]: InactiveTasks

File: api/src/main/java/io/druid/timeline/TaskDataSegment.java [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+    private final String type;
+    private final String id;
+    private final String groupId;
+    private final String dataSource;
+    private final Map<String, Object> ioConfig;
+    private final Interval interval;
+
+    @JsonCreator
+    public TaskDataSegment(
```
Comment (fjy) [inline api/src/main/java/io/druid/timeline/TaskDataSegment.java]: Are we sure this class is needed? What is the class used for serde when writing to pending Segments table?

**Comment by fjy** [comment]
@gianm is there any possible race condition where active=0 but we still need the segment?

**Comment by KurtYoung** [comment]
@fjy Yeah, i will take a look today

**Comment by haoxiang47** [comment]
Yeah thanks for review, I will keep the code style first and update my code, so excepet the code style, if there has any other problem? My logic will simplely like this:

compare the druid_tasks and druid_pendingSegments, find the dataset which 'active' is 0 in druid_tasks and match the sequence_name in druid_pendingSegments and baseSequenceName in druid_tasks's payload ioConfig parameter which type is kafka.

and then I have the question that where should this interface to action? maybe in overlord or kafka supervisior?

**Comment by jon-wei** [comment]
Closing this since #3565 was addressed by #5149

**Pull request closed without merging**